Sources/_StringProcessing/Engine/Processor.swift

```diff
@@ -703,3 +687,86 @@ extension Processor {
     }
   }
 }
+
+extension String {
+
+  func match(
+    _ char: Character,
+    at pos: Index,
+    limitedBy end: String.Index
+  ) -> Index? {
+    // TODO: This can be greatly sped up with string internals
+    // TODO: This is also very much quick-check-able
+    assert(end <= endIndex)
+
+    guard pos < end, self[pos] == char else { return nil }
+
+    let idx = index(after: pos)
+    guard idx <= end else { return nil }
+
+    return idx
+  }
+
+  func matchSeq(
+    _ seq: Substring,
+    at pos: Index,
+    limitedBy end: Index
+  ) -> Index? {
+    // TODO: This can be greatly sped up with string internals
+    // TODO: This is also very much quick-check-able
+    assert(end <= endIndex)
+
+    var cur = pos
+    for e in seq {
+      guard cur < end, self[cur] == e else { return nil }
+      self.formIndex(after: &cur)
+    }
+
+    guard cur <= end else { return nil }
+    return cur
+  }
+
+  func matchScalar(
+    _ scalar: Unicode.Scalar,
+    at pos: Index,
+    limitedBy end: String.Index,
+    boundaryCheck: Bool
+  ) -> Index? {
+    assert(end <= endIndex)
+
+    guard pos < end, unicodeScalars[pos] == scalar else {
+      return nil
+    }
+
+    let idx = unicodeScalars.index(after: pos)
+    guard idx <= end else { return nil }
+
+    if boundaryCheck && !isOnGraphemeClusterBoundary(idx) {
+      return nil
+    }
+
+    return idx
+  }
+
+  func matchBitset(
+    _ bitset: DSLTree.CustomCharacterClass.AsciiBitset,
+    at pos: Index,
+    limitedBy end: Index
+  ) -> Index? {
+    // TODO: extremely quick-check-able
+    // TODO: can be sped up with string internals
+
+    assert(end <= endIndex)
+
+    guard pos < end, bitset.matches(char: self[pos]) else {
+      return nil
+    }
+
+    let idx = index(after: pos)
```

Comment:
Scalar semantics seem to be implicitly supported here, since we're only going to match in the ASCII domain, but what about an inverted CCC? IIRC `[^abc]` has an ASCII bitset representation, but would match only the first scalar in a multi-scalar character in scalar mode.

Can you add a `FIXME` and I'll circle back to investigate?